src/Kiota.Builder/Validation/UnsupportedInheritanceSemantics.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Kiota.Builder.Configuration;
+using Kiota.Builder.Extensions;
+using Microsoft.OpenApi.Models;
+using Microsoft.OpenApi.Validations;
+
+namespace Kiota.Builder.Validation;
+public class UnsupportedInheritanceSemantics : ValidationRule<OpenApiDocument>
+{
+    public UnsupportedInheritanceSemantics(GenerationConfiguration configuration) : base((context, document) =>
+    {
+        if (document.Components != null)
+            Parallel.ForEach(document.Components.Schemas, entry =>
+            {
+                ValidateSchema(entry.Value, context);
+            });
+        var inlineSchemasToValidate = document.Paths
+                                ?.SelectMany(static x => x.Value.Operations.Values.Select(y => (x.Key, Operation: y)))
+                                .SelectMany(x => x.Operation.GetResponseSchemas(OpenApiOperationExtensions.SuccessCodes, configuration.StructuredMimeTypes).Select(static y => y))
+                                .Where(static x => string.IsNullOrEmpty(x.Reference?.Id))
+                                .ToArray() ?? Array.Empty<OpenApiSchema>();
+        Parallel.ForEach(inlineSchemasToValidate, entry =>
+        {
+            ValidateSchema(entry, context);
+        });
+    })
+    {
+    }
+
+    private static IEnumerable<string> GetAllProperties(string prefix, OpenApiSchema schema)
+    {
+        var fullPrefix = string.IsNullOrEmpty(prefix) ? "" : prefix + ".";
+        var inlinedProperties = schema.Properties
+                .Concat(schema.AllOf.SelectMany(static x => x.Properties))
+                .Concat(schema.AnyOf.SelectMany(static x => x.Properties))
+                .Concat(schema.OneOf.SelectMany(static x => x.Properties));
+
+        var currentProps = inlinedProperties.Select(p => fullPrefix + p.Key);
+        var nestedProps = inlinedProperties.SelectMany(p => GetAllProperties(fullPrefix + p.Key, p.Value));
+        return currentProps.Concat(nestedProps);
+    }
+    private static void ValidateSchema(OpenApiSchema schema, IValidationContext context)
+    {
+        var allProperties = GetAllProperties("", schema);
+        if (allProperties.Count() != allProperties.Distinct().Count())
```

Comment:
This kind of comparison will not work for the case when a subtype redefines a property simply to override the default value. I think the distinct should probably using a comparer implementation (check the overload that accepts a single argument) so properties are only considered different when they have the same name but a different type/format.

Reply:
> so properties are only considered different when they have the same name but a different type/format.

I was thinking about this as well, but, at the same time, "overriding" even only the default value is "undefined behavior" from the spec POV. Emitting a warning we are always on the safe side.

I'm mostly thinking of the gradual evolution of a specification, the override of a default value in version 1 of a spec can easily become a type + default value in version 2 as the syntax easily allows those changes.

That said, I'm not fussy about this, and if you are really convinced about it I can improve the validator 👍

Reply:
We (Microsoft Graph) are making use of default value override. This new rule would trigger a lot of "false positive" warnings. So if you don't mind I'd like the validation rule to be updated to account for that case before it gets merged, which we can always make evolve in the future.

Reply:
ok, deal 👍

Reply:
@baywet I quickly found out that expanding the Type/Format in the Validator seems to be a pretty expensive computation ...
I found `OpenApiSchemaComparer` but it takes `Default` into consideration, do we already have a generic implementation of a function `OpenApiSchema.Type => CodeType` ? Can you, please, point me in the right direction?

Reply:
yeah crawling all parent schemas for each schema is going to be expensive o(n2) at least.
In the builder I'm using an inheritance index, maybe you could use that instead? https://github.com/microsoft/kiota/blob/b593b5ddc4b6b34e69bc32540991585716345b99/src/Kiota.Builder/KiotaBuilder.cs#L1493

Reply:
Maybe I'm missing something, but, in this case, even using an inheritance index I would still need to recursively expand all of the types to check the shape.

The problem here is that Kiota is supporting default values overload which is not supposed to be supported 🙂.
I see a few possible outcomes:

- we integrate the warning as is - and you disable the rule when running on Microsoft Graph
- we track inheritance as a future improvement to this codebase and we define more precise semantics for handling it
- I drop this PR and we keep going until we gather more feedback from the community

open to other alternatives too!

Reply:
The main difference is the index is built once, when in the current PR, we end up crawling multiple times base types that have multiple derived types, multiple times in the GetAllProperties method. I think by fixing that, we could end up with o(n) instead of o(n2). May I suggest you tinker with this approach and share results before we make any decision?

Reply:
@baywet I failed to effectively use the proposed `InheritanceIndex` 🙁 , but converted the implementation to make use of a custom comparator that might be "good enough".
Let me know what you think about this proposal.

Reply:
(I'll add the changelog and documentation if we reach an agreement on the implementation)